### PR TITLE
Add Spawner Potential API

### LINKED ISSUE #1611: MobSpawner API improvements
I have some suggestions to improve the spawner API, since there are missing some fundamental features.

1) 
Currently it is only possible to set a EntityType, but via command blocks you are able to set itemstacks.
So my first suggestion: 
Allow ItemStacks to be set as a spawn result for spawners.

Following command can be used to spawn a spawner that can spawn ItemStacks (Items)

`/setblock ~1 ~1 ~1 minecraft:spawner{SpawnCount:4,SpawnRange:32,Delay:20,MinSpawnDelay:60,MaxSpawnDelay:100,RequiredPlayerRange:12,SpawnData:{id:"minecraft:item",Item:{id:"minecraft:diamond",Count:5b,tag:{display:{Name:"{\"text\":\"Super Diamond\",\"color\":\"aqua\"}"}}}}} replace`

2)
It is also possible to set a list of "SpawnPotentials" with a defined "Weight" so spawners can spawn random entities from a list with a defined chance. But also this part is missing in the current API. 

Example Command to get this kind of spawner:
`/setblock ~1 ~1 ~1 minecraft:spawner{SpawnCount:4,SpawnRange:32,Delay:20,MinSpawnDelay:60,MaxSpawnDelay:100,RequiredPlayerRange:12,SpawnData:{id:"minecraft:chicken"},SpawnPotentials:[{Weight:50,Entity:{id:"minecraft:chicken"}},{Weight:40,Entity:{id:"minecraft:cow"}},{Weight:3,Entity:{id:"minecraft:bat"}}]} replace`

### PR DESCRIPTION
**Pull request by tahmid-23**
Does part 2 of and fixes #1611

### CODE REVIEW

File: patches/api/0324-Add-Spawner-Potential-API.patch [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+-public interface CreatureSpawner extends TileState {
++public interface CreatureSpawner extends TileState, io.papermc.paper.world.Spawner { // Paper - extend Spawner
+ 
+     /**
+      * Get the spawner's creature type.
+@@ -16,8 +16,10 @@ public interface CreatureSpawner extends TileState {
+     @NotNull
+     public EntityType getSpawnedType();
+ 
++    // Paper - update javadoc
```
Comment (Machine-Maker) [inline patches/api/0324-Add-Spawner-Potential-API.patch]: Don't need to comment javadoc changes

File: patches/api/0324-Add-Spawner-Potential-API.patch [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+      * @see #setSpawnedType(EntityType)
+diff --git a/src/main/java/org/bukkit/entity/minecart/SpawnerMinecart.java b/src/main/java/org/bukkit/entity/minecart/SpawnerMinecart.java
+index 0ce3592ec7b121413bb2203f3d6056e450b89736..b4f33aefa5dff2b21ed46d33511a2328a4e2933f 100644
+--- a/src/main/java/org/bukkit/entity/minecart/SpawnerMinecart.java
++++ b/src/main/java/org/bukkit/entity/minecart/SpawnerMinecart.java
+@@ -2,9 +2,9 @@ package org.bukkit.entity.minecart;
+ 
+ import org.bukkit.entity.Minecart;
+ 
++// Paper start - update javadoc
```
Comment (Machine-Maker) [inline patches/api/0324-Add-Spawner-Potential-API.patch]: another unneeded comment

File: patches/server/0728-Implement-Spawner-Potential-API.patch [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+index c601b8b12756682a4cb300be8ebed4319902c5b5..00ff06c27c4f04a094135d627515f400a0e3f857 100644
+--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
++++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+@@ -361,6 +361,14 @@ public abstract class BaseSpawner {
+         this.nextSpawnData = mobspawnerdata;
+     }
+ 
++    // Paper start
```
Comment (Machine-Maker) [inline patches/server/0728-Implement-Spawner-Potential-API.patch]: If this logic in this function is copied from somewhere else in `BaseSpawner`, make a note of where its taken from for future updaters of this patch.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
There is still some missing stuff in here for spawners, for example entities can have equipment, or anything really but that's very hard to do with bukkit's current Entity API.

**Comment by tahmid-23** [comment]
rebased + did suggestions

**Comment by tahmid-23** [comment]
not quite stale

**Comment by tahmid-23** [comment]
shoo + would like suggestion on a better way with itemstacks vs entities for spawner entries

**Pull request closed without merging**